Alex Blinov dz2

obscene3190 opened the pull request.

obscene3190 committed:
added dz

obscene3190 committed:
logging constants

obscene3190 committed:
added fh closing

obscene3190 committed:
back to origin

obscene3190 committed:
updated destructor

obscene3190 committed:
logs fix(Ivan)

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+def check_date(data_type, data):
+    # first name
+    if data_type == 'fname':
+        if data is int:
```
Явно имелось в виду isinstance(data, int). Текущее условие не выполнится практически никогда

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+        if data is int:
+            raise TypeError('Incorrect first name type')
+        if (re.search(r'\d', data)) is None:
+            if re.match(r'\w+\b', data):
```
Можно оставить только одно это условие

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+            raise ValueError('Incorrect first name value')
+    # second name
+    if data_type == 'sname':
+        if data is int:
```
см. коммент выше

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+    if data_type == 'sname':
+        if data is int:
+            raise TypeError('Incorrect second name type')
+        if (re.search(r'\d', data)) is None:
```
код дублируется - можно схлопнуть fname и sname в один блок

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+    # birth date
+    if data_type == 'bdate':
+        data = re.split('-', data)
+        data = ''.join(data)
```
тогда уж data.replace('-', '')

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+        data = ''.join(data)
+        if re.search(r'\D', data) is None:
+            data = re.findall(r'\d+', data)
+            data = ''.join(data)
```
вот это всё проще сделать через что-то вроде re.match с регуляркой "\d{4}-\d{2}-\d{2}"

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+        types = ('Паспорт', 'паспорт', 'Заграничный паспорт',
+                 'заграничный паспорт', 'загран',
+                 'Водительские права', 'водительсикие права')
+        if (re.search(r'\d', data)) is None:
```
кажется, что много лишних условий

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+        if re.search(r'[^\d ]]', data) is None:
+            data = re.findall(r'\d+', data)
+            data = ''.join(data)
+            if len(data) == 10 or len(data) == 9:
```
здесь пройдет паспорт с неправильной длиной

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+            raise ValueError('Incorrect doc num value')
+
+
+def Logger(func):
```
функции именуются маленькими буквами

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+        try:
+            func(self, *args, *kwargs)
+        except TypeError:
+            self.ex_log.error('Incorrect input type')
```
в таком виде лог крайне неинформативен. Нужно добавить больше информации - данные о пациенте и неправильные данные

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+        self.log = logging.getLogger('Logger')
+
+        try:
+            self.__first_name = check_date('fname', first_name)
```
зачем двойное подчеркивание?

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+            self.document_type_ = check_date('dtype', document_type)
+            self.document_id_ = check_date('docid', document_id)
+        except TypeError as error_text:
+            self.ex_log.error('Incorrect input value')
```
неинформативно

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+                          (first_name, second_name, birth_date, phone, document_type, document_id))
+
+    @classmethod
+    def create(cls, *args):
```
здесь не используются никакие атрибуты класса, стоит сделать этот метод статическим

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+                patient = [self.__first_name, self.__last_name, self.birth_date_,
+                           self.phone_, self.document_type_, self.document_id_]
+                writer.writerow(patient)
+        except Exception:
```
слишком общий except

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+    @birth_date.setter
+    @Logger
+    def birth_date(self, value):
+        self.birth_date_ = check_date('bdate', value)
```
было бы лучше разнести логику по проверке разных по смыслу значений в разные функции (а в данном случае - непосредственно в сеттеры), потому что зачем в одном огромном декораторе повторять все те же сущности, что представлены в атрибутах класса

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+        return self.phone_
+
+    @phone.setter
+    @Logger
```
не думаю, что тут была необходимость в декораторе. Можно просто функцию написать, которая вызывает ошибки. Простое всегда лучше, чем сложное)

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+        self.document_id_ = check_date('docid', value)
+
+    def __del__(self):
+        del self.log, self.ex_log
```
это и так произойдет

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
+
+    def __del__(self):
+        del self.log, self.ex_log
+        for fh in list(logging.getLogger('Exception_Logger').handlers)[::-1]:
```
с учетом последних замечаний, нам не нужно, чтобы все хэндлеры закрывались при уничтожении какого-то одного единственного объекта

romvano commented on homework/patient.py:
```diff
@@ -1,17 +1,228 @@
-    def limit(self, n):
-        raise NotImplementedError()
+    def __del__(self):
+        del self.log, self.ex_log
```
те же замечания, что в классе Patient

romvano reviewed (commented):
В целом: подход запихать все проверки в одну функцию - не торт. Логика должна быть читаемой. Много лишних условий в регулярках. Неверное применение "private"-нотации.

Не увидел корректную обработку TypeError.

Декоратор считаю излишним усложнением, но это не столь критично.

obscene3190 committed:
исправление замечаний

проверка значений перенесена в сеттеры
создана маленькая функция для логирования и выкидывания исключений

obscene3190 committed:
idk why

obscene3190 committed:
- private

obscene3190 committed:
__- -___ -_-___- -_____ 

________________----_______-_---

obscene3190 committed:
Create code.c
